File: src/service/cmds.js
import { invoke } from "@tauri-apps/api/tauri";
import { sendNotice } from "@/service/msg";

export async function getCommonConfig() {
  return invoke("get_common_config");
}

export async function setCommonConfig(config) {
  return invoke("set_common_config", { config });
}

export async function setLanguage(language) {
  return invoke("change_language", { language });
}

export async function setRecordLimit(limit) {
  return invoke("change_record_limit", { limit });
}

export async function setAutoLaunch(enable) {
  return invoke("change_auto_launch", { enable });
}

export async function setThemeMode(themeMode) {
  return invoke("change_theme_mode", { themeMode });
}

export async function setHotkeys(hotkeys) {
  return invoke("change_hotkeys", { hotkeys });
}

export async function clearData() {
  return invoke("clear_data");
}

export async function insertRecord(record) {
  return invoke("insert_record", { record });
}

export async function insertIfNotExist(r) {
  return invoke("insert_if_not_exist", { r });
}

export async function findAllRecord() {
  return invoke("find_all_record");
}

export async function markFavorite(id) {
  return invoke("mark_favorite", { id });
}

export async function saveTags(id, tags) {
  return invoke("save_tags", { id, tags: tags.join(",") });
}

export async function findByKey(query) {
  return invoke("find_by_key", { query });
}

export async function deleteOverLimit(limit) {
  return invoke("delete_over_limit", { limit });
}

export async function writeToClip(id) {
  invoke("write_to_clip", { id });
  sendNotice("", "Copy!");
}

export async function deleteById(id) {
  return invoke("delete_by_id", { id });
}

File: src-tauri/src/cmds.rs
use crate::{
    config,
    config::{CommonConfig, Config},
    core::{
        clipboard::{ClipBoardOprator, ImageDataDB},
        database::{QueryReq, Record, SqliteDB},
        handle::Handle,
    },
    log_err,
    utils::json_util,
};

type CmdResult<T = ()> = Result<T, String>;

#[tauri::command]
pub fn get_common_config() -> CmdResult<CommonConfig> {
    Ok(Config::common().data().clone())
}

#[tauri::command]
pub fn set_common_config(config: CommonConfig) -> CmdResult {
    Config::common().draft().patch_config(config);
    Config::common().apply();
    log_err!(Config::common().data().save_file());

    // todo enable_auto_launch
    // todo hotkeys
    Handle::refresh_common_config();
    log_err!(Handle::update_systray());
    Ok(())
}

#[tauri::command]
pub async fn change_language(language: String) -> CmdResult {
    let _ = config::modify_common_config(CommonConfig {
        language: Some(language),
        ..CommonConfig::default()
    })
    .await;
    Ok(())
}

#[tauri::command]
pub async fn change_record_limit(limit: u32) -> CmdResult {
    let _ = config::modify_common_config(CommonConfig {
        record_limit: Some(limit),
        ..CommonConfig::default()
    })
    .await;
    Ok(())
}

#[tauri::command]
pub async fn change_auto_launch(enable: bool) -> CmdResult {
    let _ = config::modify_common_config(CommonConfig {
        enable_auto_launch: Some(enable),
        ..CommonConfig::default()
    })
    .await;
    Ok(())
}

#[tauri::command]
pub async fn change_theme_mode(theme_mode: String) -> CmdResult {
    let _ = config::modify_common_config(CommonConfig {
        theme_mode: Some(theme_mode),
        ..CommonConfig::default()
    })
    .await;
    Ok(())
}

#[tauri::command]
pub async fn change_hotkeys(hotkeys: Vec<String>) -> CmdResult {
    let _ = config::modify_common_config(CommonConfig {
        hotkeys: Some(hotkeys),
        ..CommonConfig::default()
    })
    .await;
    Ok(())
}

#[tauri::command]
pub fn clear_data() -> bool {
    match SqliteDB::new().clear_data() {
        Ok(()) => true,
        Err(e) => {
            println!("{}", format!("清除失败:{}", e));
            false
        }
    }
}

#[tauri::command]
pub fn insert_record(r: Record) -> bool {
    match SqliteDB::new().insert_record(r) {
        Ok(_i) => true,
        Err(e) => {
            println!("err:{}", e);
            false
        }
    }
}

#[tauri::command]
pub fn insert_if_not_exist(r: Record) -> bool {
    match SqliteDB::new().insert_if_not_exist(r) {
        Ok(_i) => true,
        Err(e) => {
            println!("err:{}", e);
            false
        }
    }
}

#[tauri::command]
pub fn find_all_record() -> Vec<Record> {
    SqliteDB::new().find_all().unwrap()
}

#[tauri::command]
pub fn mark_favorite(id: u64) -> bool {
    match SqliteDB::new().mark_favorite(id) {
        Ok(_i) => true,
        Err(e) => {
            println!("err:{}", e);
            false
        }
    }
}

#[tauri::command]
pub fn delete_by_id(id: u64) -> bool {
    match SqliteDB::new().delete_by_id(id) {
        Ok(_i) => true,
        Err(e) => {
            println!("err:{}", e);
            false
        }
    }
}

#[tauri::command]
pub fn find_by_key(query: QueryReq) -> Vec<Record> {
    SqliteDB::new().find_by_key(query).unwrap()
}

#[tauri::command]
pub fn delete_over_limit(limit: usize) -> bool {
    match SqliteDB::new().delete_over_limit(limit) {
        Ok(_i) => true,
        Err(e) => {
            println!("err:{}", e);
            false
        }
    }
}

#[tauri::command]
pub fn write_to_clip(id: u64) -> bool {
    let record = SqliteDB::new().find_by_id(id);
    match record {
        Ok(r) => {
            if r.data_type == "text" {
                let _ = ClipBoardOprator::set_text(r.content);
            } else if r.data_type == "image" {
                let image_data: ImageDataDB = json_util::parse(&r.content).unwrap();
                let _ = ClipBoardOprator::set_image(image_data);
            }
            true
        }
        Err(e) => {
            println!("err:{}", e);
            false
        }
    }
}
